ts/session/sending/PendingMessageCache.ts

```diff
@@ -1,36 +1,120 @@
 
   constructor() {
-    // TODO: We should load pending messages from db here
+    // Load pending messages from the database
+    // You must call init() on this class in order to load from DB.
+    //    const pendingMessageCache = new PendingMessageCache();
+    //    await pendingMessageCache.init()
+    //    >> do stuff
+    this.cache = [];
+  }
+
+  public async add(
+    device: PubKey,
+    message: ContentMessage
+  ): Promise<RawMessage> {
+    const rawMessage = MessageUtils.toRawMessage(device, message);
+
+    // Does it exist in cache already?
+    if (this.find(rawMessage)) {
+      return rawMessage;
+    }
+
+    this.cache.push(rawMessage);
+    await this.syncCacheWithDB();
+
+    return rawMessage;
+  }
+
+  public async remove(
+    message: RawMessage
+  ): Promise<Array<RawMessage> | undefined> {
+    // Should only be called after message is processed
+
+    // Return if message doesn't exist in cache
+    if (!this.find(message)) {
+      return;
+    }
+
+    // Remove item from cache and sync with database
+    const updatedCache = this.cache.filter(
+      m => m.identifier !== message.identifier
+    );
+    this.cache = updatedCache;
+    await this.syncCacheWithDB();
+
+    return updatedCache;
   }
 
-  public addPendingMessage(
-    device: string,
-    message: OutgoingContentMessage
-  ): RawMessage {
-    // TODO: Maybe have a util for converting OutgoingContentMessage to RawMessage?
-    // TODO: Raw message has uuid, how are we going to set that? maybe use a different identifier?
-    // One could be device + timestamp would make a unique identifier
-    // TODO: Return previous pending message if it exists
-    return {} as RawMessage;
+  public find(message: RawMessage): RawMessage | undefined {
+    // Find a message in the cache
+    return this.cache.find(
+      m => m.device === message.device && m.timestamp === message.timestamp
+    );
   }
 
-  public removePendingMessage(message: RawMessage) {
-    // TODO: implement
+  public getForDevice(device: PubKey): Array<RawMessage> {
+    return this.cache.filter(m => m.device === device.key);
   }
 
-  public getPendingDevices(): Array<String> {
-    // TODO: this should return all devices which have pending messages
-    return [];
+  public async clear() {
+    // Clears the cache and syncs to DB
+    this.cache = [];
+    await this.syncCacheWithDB();
+  }
+
+  public getDevices(): Array<PubKey> {
+    // Gets all devices with pending messages
+    const pubkeys = [...new Set(this.cache.map(m => m.device))];
+
+    return pubkeys.map(d => PubKey.from(d));
+  }
+
+  public async init() {
+    const messages = await this.getFromStorage();
+    this.cache = messages;
+  }
+
+  public async getFromStorage(): Promise<Array<RawMessage>> {
+    // tslint:disable-next-line: no-backbone-get-set-outside-model
+    const pendingMessagesData = await Data.getItemById('pendingMessages');
+    const pendingMessagesJSON = pendingMessagesData
```

Comment:
you might be able to simplify this to be more readable
```js
const data = await Data.getItemById('pendingMessages');
if (!data || !data.value) { return []; }

const json = JSON.parse(String(data.value));
// Do other stuff
return json;
```

Reply:
Done